### ISSUE TITLE
Migrating from SoftHSM 1 to 2: "ERROR: Could not  get the class of object 1. Continuing."

### ISSUE THREAD

**Issue by bortzmeyer** [opendnssec/SoftHSMv2#368]
I want to migrate an old installation, which still uses SoftHSMv1.

On the new machine, I initialized SoftHSM v2:

```
sudo softhsm2-util --init-token  --slot 0 --label OpenDNSSEC
```

But then migration fails:

```
% sudo softhsm2-migrate --db /var/lib/softhsm/slot0.db --token OpenDNSSEC
Found slot 1877891147 with matching token label.
=== User PIN (4-255 characters) ===
Please enter user PIN: ********************************
Please reenter user PIN: ********************************
ERROR: Could not get the class of object 1. Continuing.
ERROR: Could not get the class of object 2. Continuing.
ERROR: Could not get the class of object 3. Continuing.
ERROR: Could not get the class of object 4. Continuing.
ERROR: Could not get the class of object 41. Continuing.
ERROR: Could not get the class of object 42. Continuing.
ERROR: Could not get the class of object 51. Continuing.
ERROR: Could not get the class of object 52. Continuing.
ERROR: Could not get the class of object 55. Continuing.
ERROR: Could not get the class of object 56. Continuing.
ERROR: Could not get the class of object 89. Continuing.
ERROR: Could not get the class of object 90. Continuing.
ERROR: Could not get the class of object 191. Continuing.
ERROR: Could not get the class of object 192. Continuing.
ERROR: Could not get the class of object 193. Continuing.
...
ERROR: Could not get the class of object 292. Continuing.
ERROR: Unable to migrate all of the objects.
```

**Comment by bellgrim** [comment]
Have you moved the database file from a 32-bit system to a 64-bit system?

softhsm2-migrate will check if the length matches the sizeof(CK_OBJECT_CLASS)): unsigned long. The tool does not support converting from 32-bit to 64-bit.

It is 4 or 8 bytes depending on the system.

```
$ sqlite3 /var/lib/softhsm/slot0.db 'SELECT length FROM Attributes WHERE type=0;'
8
8
```

**Comment by bortzmeyer** [comment]
@bellgrim Indeed, this is the case.

```
% sqlite3 /var/lib/softhsm/slot0.db 'SELECT length FROM Attributes WHERE type=0;'
4
4
...
```

And the new machine is 64 bits. Does it mean there is no way?

**Comment by bellgrim** [comment]
What we could do is to have a list of attributes that are stored as CK_ULONG and allow both 4 and 8 bytes on 64-bit systems when reading from SoftHSMv1 databases.

**Comment by rijswijk** [comment]
Schedule this for 2.5.0 and ask @bortzmeyer if he is willing to test.

**Comment by bellgrim** [comment]
Should be fixed in #380 

@bortzmeyer Does this work for you?
